src/Client/Connection.cpp

```diff
@@ -405,6 +442,40 @@ void Connection::sendQuery(
     else
         writeStringBinary("" /* empty string is a marker of the end of settings */, *out);
 
+    /// Interserver secret
+    if (server_revision >= DBMS_MIN_REVISION_WITH_INTERSERVER_SECRET)
+    {
+        /// Hash
+        ///
+        /// Send correct hash only for !INITIAL_QUERY, due to:
+        /// - this will avoid extra protocol complexity for simplest cases
+        /// - there is no need in hash for the INITIAL_QUERY anyway
+        ///   (since there is no secure/unsecure changes)
+        if (client_info && !cluster_secret.empty() && client_info->query_kind != ClientInfo::QueryKind::INITIAL_QUERY)
+        {
+#if USE_SSL
+            std::string data(salt);
+            data += cluster_secret;
+            data += query;
+            data += query_id;
+            data += user;
+            data += client_info->initial_user;
```

Comment:
Why are you sending `user` here? It seems at this point `user` is already sent in the hello package, and it cannot change like `initial_user`.

Reply:
Just to make hash depends on both users:
- initial
- and from the hello package

To decrease probability of hijacking as much as possible.

Reply:
User from the hello package is the "remote" user defined in the cluster configuration. If the interserver secret is enabled there is no much sense to keep this users in the cluster configuration. So `user` here will be probably an empty string and so not very useful to prevent hijacking.

Reply:
Ok, kind of agree.

Reply:
Further, you could kind of add more security because you can get the "remote" user from the cluster configuration locally instead of transferring it between nodes. However actually it wouldn't add more security because in this case that user name would work exactly as the second interserver secret.